vyperlang/vyper
# Type checking: draft for comments, round 2

### What I did
Add a type checking pass.

This is round two of #1918, featuring less complexity :sunglasses: 

### How I did it
Added the `vyper/context` subpackage. See `vyper/context/README.md` and various docstrings.

### How to verify it
I have not included updates to tests in order to keep this diff smaller (they add another 2000 lines).  I have another branch [`type-check-x`](https://github.com/vyperlang/vyper/compare/master...iamdefinitelyahuman:type-check-x) where all tests are passing.

### TODO
* [ ] code review, feel good about the implementation
* [ ] finish implementing `mypy` within the new code (this is the only job currently failing)
* [ ] finish logic for `convert` builtin
* [ ] remove `raise` statements from `vyper/parser` - any that are still green in the coverage report are an indication of some missing logic in the type check pass
* [ ] plan and then write new test cases for type checking, purge some old tests

### Cute Animal Picture
![image](https://user-images.githubusercontent.com/35276322/85035477-48084880-b194-11ea-82b6-6d64ed66113d.png)
